packages/hub/services/scheduled-payments/executor.ts

```diff
@@ -51,6 +52,25 @@ export default class ScheduledPaymentsExecutorService {
     return gasPriceInToken(provider, gasTokenAddress);
   }
 
+  async executeScheduledPayments() {
+    // TODO: take a only a few, can be 10, to not overwhelm this method
+    // TODO: randomize the order
+    let scheduledPayments = await this.scheduledPaymentFetcher.fetchScheduledPayments();
+
+    // Currently we do one by one, but we can do in parallel when we'll have a lot of scheduled payments
```

Comment:
actually i wonder about this--the crank's EOA nonce could get out of sync if we do parallel on chain calls. The relay server actually uses redis to obtain a lock on the EOA as part of executing a txn on chain specifically so the nonce associated with the EOA will not get out of sync. I think eventually we will probably want to employ something similar to make sure that on chain calls are always performed sequentially.

Reply:
Here you can see an example of this--any txn sent from the relay EOA must use an `EthereumNonceLock` (which is backed by redis in this case)

Reply:
the danger here is that if you have a nonce collision (because 2 concurrent tasks use the same nonce), only one of the txns will be considered valid--and it will be up to the miner that makes the block that determines this. all the other txn with the same nonce will be invalid since the ethereum protocol does not allow mulitple txn's for the same address to have the same nonce

Reply:
@habdelra is it possible to add Redis for the hub? to implement this solution?

Reply:
we could also just use the DB we already have. either redis or a postgres DB could be used to facilitate a locking mechanism. I would think using the existing postgres DB is probably a lot easier since we alread have that service configured in all our envs.

Reply:
Alright, I'll use postgres DB

Reply:
Agreed on using postgres with a transaction for this.

Reply:
Could PG's [advisory locks](https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-ADVISORY-LOCKS) come handy here? They are useful for application defined locking

Reply:
I have added the nonce lock solution

Reply:
I've used advisory locks in the past, and yes, they are created for situations like these. in this case there only ever is one crank EOA, so i think a single lock is all you need (as opposed to a solution that had multiple crank EOA's in which case each crank would need a lock)

Reply:
updated, used advisory locks instead of pessimistic locks